File: Common/c3270/mkkeypad.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/*
 * Construct keypad data structures from a set of descriptor files.
 *
 * The files are:
 *  keypad.labels
 *   literal text to be drawn for the keypad
 *  keypad.outline
 *   outlines for the keys, ACS encoded ('l' for upper left, etc.)
 *  keypad.map
 *   sensitivity map for the keypad (aaaa is field 'a', etc.)
 *  keypad.full
 *   not used by this program, but gives the overall plan
 *
 * The result is an array of structures:
 *  unsigned char literal;	text from keypad.labels
 *  unsigned char outline;	ACS-encoded outline text
 *  sens_t *sens;		sensitivity, or NULL
 *
 * A sens_t is a structure:
 *  unsigned char ul_x, ul_y;	upper left corner
 *  unsigned char lr_x, lr_y;	lower right corner
 *  unsigned char callback_name; 'a', 'b', etc.
 */

typedef struct sensmap {
    struct sensmap *next;
    unsigned char name;
    unsigned ul_x, ul_y, lr_x, lr_y;
    int index;
    char *callback;
} sensmap_t;

sensmap_t *sensmaps = NULL;
sensmap_t *last_sensmap = NULL;
int sensmap_count = 0;

char *incdir = NULL;

FILE *fopen_inc(const char *name)
{
    FILE *f;
    char *path;

    if ((f = fopen(name, "r")) != NULL) {
	return f;
    }
    if (incdir == NULL) {
	return NULL;
    }
    path = malloc(strlen(incdir) + 1 + strlen(name) + 1);
    sprintf(path, "%s/%s", incdir, name);
    f = fopen(path, "r");
    free(path);
    return f;
}

int
main(int argc, char *argv[])
{
    FILE *callbacks;
    FILE *labels;
    FILE *outline;
    FILE *map;
    int c, d;
    unsigned x;
    unsigned y;
    sensmap_t *s;
    char buf[128];
    int cbl = 0;

    if (argc > 1 && !strncmp(argv[1], "-I", 2)) {
	incdir = argv[1] + 2;
    }

    /* Open the files. */
    labels = fopen_inc("keypad.labels");
    if (labels == NULL) {
	perror("keypad.labels");
	exit(1);
    }
    outline = fopen_inc("keypad.outline");
    if (outline == NULL) {
	perror("keypad.outline");
	exit(1);
    }
    map = fopen_inc("keypad.map");
    if (map == NULL) {
	perror("keypad.map");
	exit(1);
    }
    callbacks = fopen_inc("keypad.callbacks");
    if (callbacks == NULL) {
	perror("keypad.callbacks");
	exit(1);
    }

    /* Read in the map file first. */
    x = 0;
    y = 0;
    while ((c = fgetc(map)) != EOF) {
	if (c == '\n') {
	    y++;
	    x = 0;
	    continue;
	}
	if (c == ' ') {
	    x++;
	    continue;
	}
	for (s = sensmaps; s != NULL; s = s->next) {
	    if (s->name == c)
		break;
	}
	if (s != NULL) {
	    /* Seen it before. */
	    s->lr_x = x;
	    s->lr_y = y;
	} else {
	    s = (sensmap_t *)malloc(sizeof(sensmap_t));
	    if (s == NULL) {
		fprintf(stderr, "Out of memory.\n");
		exit(1);
	    }
	    memset(s, '\0', sizeof(sensmap_t));
	    s->name = c;
	    s->ul_x = s->lr_x = x;
	    s->ul_y = s->lr_y = y;
	    s->index = sensmap_count++;
	    s->callback = NULL;
	    s->next = NULL;
	    if (last_sensmap)
		last_sensmap->next = s;
	    else
		sensmaps = s;
	    last_sensmap = s;
	    last_sensmap = s;
	}
	x++;
    }
    fclose(map);

    /* Read in the callbacks. */
    while (fgets(buf, sizeof(buf), callbacks) != NULL) {
	char *t;
	char c;
	size_t sl;

	cbl++;
	sl = strlen(buf);
	if (sl > 0 && buf[sl - 1] == '\n')
	    buf[sl - 1] = '\0';
	c = buf[0];
	if (!isalnum((unsigned char)c)) {
	    fprintf(stderr, "keypad.callbacks:%d Invalid callback character.\n",
		    cbl);
	    exit(1);
	}
	t = &buf[1];
	while (*t && isspace((unsigned char)*t)) {
	    t++;
	}
	if (!*t || !isalnum((unsigned char)*t)) {
	    fprintf(stderr, "keypad.callbacks:%d Invalid callback string.\n",
		    cbl);
	    exit(1);
	}
#if defined(MKK_DEBUG) /*[*/
	fprintf(stderr, "line %d: name '%c', callback '%s'\n", cbl, c, t);
#endif /*]*/
	for (s = sensmaps; s != NULL; s = s->next) {
	    if (s->name == c) {
		if (s->callback != NULL) {
		    fprintf(stderr, "keypad.callbacks:%d Duplicate callback "
			    "for '%c' (%s, %s).\n", cbl, c,
			    s->callback, t);
		    exit(1);
		}
		s->callback = malloc(strlen(t) + 1);
		if (s->callback == NULL) {
		    fprintf(stderr, "Out of memory.\n");
		    exit(1);
		}
		strcpy(s->callback, t);
		break;
	    }
	}
	if (s == NULL) {
	    fprintf(stderr, "keypad.callbacks:%d: Callback '%c' for "
		    "nonexistent map.\n", cbl, c);
	    exit(1);
	}
    }
    fclose(callbacks);
    for (s = sensmaps; s != NULL; s = s->next) {
	if (s->callback == NULL) {
	    fprintf(stderr, "Map '%c' has no callback.\n", s->name);
	    exit(1);
	}
    }

    /* Dump out the sensmaps. */
    printf("sens_t sens[%u] = {\n", sensmap_count);
    for (s = sensmaps; s != NULL; s = s->next) {
	printf("  { %2u, %2u, %2u, %2u, \"%s\" },\n",
		s->ul_x, s->ul_y, s->lr_x, s->lr_y, s->callback);
    }
    printf("};\n");

    /*
     * Read in the label and outline files, and use them to dump out the
     * keypad_desc[].
     */
    labels = fopen_inc("keypad.labels");
    if (labels == NULL) {
	perror("keypad.labels");
	exit(1);
    }
    outline = fopen_inc("keypad.outline");
    if (outline == NULL) {
	perror("keypad.outline");
	exit(1);
    }

    printf("keypad_desc_t keypad_desc[%u][80] = {\n", y);
    printf("{ /* row 0 */\n");

    x = 0;
    y = 0;
    while ((c = fgetc(labels)) != EOF) {
	d = fgetc(outline);
	if (c == '\n') {
	    if (d != '\n') {
		fprintf(stderr, "labels and outline out of sync at line %d\n",
			y + 1);
		exit(1);
	    }
	    y++;
	    x = 0;
	    continue;
	}
	if (x == 0 && y != 0)
	    printf("},\n{ /* row %u */\n", y);
	for (s = sensmaps; s != NULL; s = s->next) {
	    if (x >= s->ul_x && y >= s->ul_y &&
		x <= s->lr_x && y <= s->lr_y) {
		printf("  { '%c', '%c', &sens[%u] },\n",
			c, d, s->index);
		break;
	    }
	}
	if (s == NULL) {
	    if (c == ' ' && d == ' ')
		printf("  {   0,   0, NULL },\n");
	    else
		printf("  { '%c', '%c', NULL },\n", c, d);
	}
	x++;
    }
    if ((d = fgetc(outline)) != EOF) {
	fprintf(stderr, "labels and outline out of sync at EOF\n");
	exit(1);
    }
    printf("} };\n");
    fclose(labels);
    fclose(outline);

    return 0;
}

File: Common/mkkeypad.py
#!/usr/bin/env python3
#
# Copyright (c) 2025 Paul Mattes.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the names of Paul Mattes nor the names of his contributors
#       may be used to endorse or promote products derived from this software
#       without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY PAUL MATTES "AS IS" AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Construct keypad data structures from a set of descriptor files.
#
# The files are:
#  keypad.labels
#   literal text to be drawn for the keypad
#  keypad.outline
#   outlines for the keys, ACS encoded ('l' for upper left, etc.)
#  keypad.map
#   sensitivity map for the keypad (aaaa is field 'a', etc.)
#  keypad.full
#   not used by this program, but gives the overall plan
#
# The result is an array of structures:
#  unsigned char literal;      text from keypad.labels
#  unsigned char outline;      ACS-encoded outline text
#  sens_t *sens;               sensitivity, or NULL
#
# A sens_t is a structure:
#  unsigned char ul_x, ul_y;   upper left corner
#  unsigned char lr_x, lr_y;   lower right corner
#  unsigned char callback_name; 'a', 'b', etc.

import os
import os.path
import sys

class Sensmap:
    '''Sensitivity map'''

    # Global index.
    index = 0

    def __init__(self, x: int, y: int):
        self.ul_x = x
        self.ul_y = y
        self.lr_x = x
        self.lr_y = y
        self.callback = None

        self.index = Sensmap.index
        Sensmap.index += 1

class fopen_inc():
    def __init__(self, name: str, incdir: str):
        self.file = None
        try:
            self.file = open(name)
        except Exception as ex:
            if incdir != None:
                self.file = open(os.path.join(incdir, name))
            else:
                raise ex

    def __enter__(self):
        return self.file

    def __exit__(self, exc_type, exc_value, exc_traceback):
        self.close()

    def close(self):
        if self.file != None:
            self.file.close()
            self.file = None

def mkkeypad(incdir: str, out_file_name: str):
    '''Generate the keypad map'''
    sensmaps = {}
    with fopen_inc('keypad.callbacks', incdir) as callbacks:
        with fopen_inc('keypad.map', incdir) as map:

            if out_file_name != None:
                out_file = open(out_file_name, 'w')
            else:
                out_file = sys.stdout

            # Read in the map file.
            x = 0
            y = 0
            while True:
                c = map.read(1)
                if c == '':
                    break
                if c == '\n':
                    y += 1
                    x = 0
                    continue
                if c == ' ':
                    x += 1
                    continue
                if c in sensmaps:
                    sensmaps[c].lr_x = x
                    sensmaps[c].lr_y = y
                else:
                    sensmaps[c] = Sensmap(x, y)
                x += 1

            # Read in the callbacks.
            while True:
                buf = callbacks.readline()
                if buf == '':
                    break
                buf = buf.strip()
                c = buf[0]
                buf = buf[1:].strip()
                if c in sensmaps:
                    sensmaps[c].callback = buf
                else:
                    print( 'Unknown map: {c}', file=sys.stderr)
                    exit(1)

    # Check the sensmaps.
    for key, s in sensmaps.items():
        if s.callback == None:
            print(f'Map "{key}" has no callback', file=sys.stderr)
            exit(1)

    # Dump out the sensmaps.
    keysort = {} # dictionary of indices to semsmaps keys
    for key, value in sensmaps.items():
        keysort[value.index] = key
    print(f'sens_t sens[{len(sensmaps)}] =', '{', file=out_file)
    for ks in sorted(keysort.keys()):
        key = keysort[ks]
        s = sensmaps[key]
        print('  { ' + f'{s.ul_x:2d}, {s.ul_y:2d}, {s.lr_x:2d}, {s.lr_y:2d}, "{s.callback}"' + ' },', file=out_file)
    print('};', file=out_file)

    # Read in the label and outline files, and use them to dump out keypad_desc[].
    with fopen_inc('keypad.labels', incdir) as labels:
        with fopen_inc('keypad.outline', incdir) as outline:

            print(f'keypad_desc_t keypad_desc[{y}][80] = ' + '{', file=out_file)
            print('{ /* row 0 */', file=out_file)
            x = 0
            y = 0
            while True:
                c = labels.read(1)
                if c == '':
                    break
                d = outline.read(1)
                if c == '\n':
                    if d != '\n':
                        print(f'labels and outline are out of sync at line {y + 1}', file=sys.stderr)
                        exit(1)
                    y += 1
                    x = 0
                    continue
                if x == 0 and y != 0:
                    print('},', file=out_file)
                    print('{ ' + f'/* row {y} */', file=out_file)
                found = False
                for _, s in sensmaps.items():
                    if x >= s.ul_x and y >= s.ul_y and x <= s.lr_x and y <= s.lr_y:
                        print('  { ' + f"'{c}', '{d}', &sens[{s.index}]" + ' },', file=out_file)
                        found = True
                        break
                if not found:
                    if c == ' ' and d == ' ':
                        print('  {   0,   0, NULL },', file=out_file)
                    else:
                        print('  { ' + f"'{c}', '{d}', NULL" + ' },', file=out_file)
                x += 1
            d = outline.read(1)
            if d != '':
                print('labels and outlines are out of sync at EOF', file=os.stderr)
                exit(1)
            print('} };', file=out_file)

    if out_file_name != None:
        out_file.close()

# Parse the command line.
incdir = None
outfile_name = None
args = sys.argv.copy()[1:]
while len(args) > 0:
    if args[0].startswith('-I'):
        incdir = args[0][2:]
    elif args[0] == '-o':
        if len(args) < 2:
            print('Missing value after -o', file=sys.stderr)
            exit(1)
        outfile_name = args[1]
        args = args[1:]
    else:
        print(f"Unknown option '{args[0]}'", file=sys.stderr)
        exit(1)
    args = args[1:]

# Run.
mkkeypad(incdir, outfile_name)